scipy/optimize/lbfgsb.py

```diff
@@ -336,9 +337,106 @@ def func_and_grad(x):
     else:
         warnflag = 2
 
+    # these two portions of the workspace are described in the mainlb
+    # subroutine in lbfgsb.f. See line 363.
+    s = wa[0: m*n].reshape(m, n)
+    y = wa[m*n: 2*m*n].reshape(m, n)
+    n_corrs = min(n_iterations-1, maxcor)
+    hess_inv = _LbfgsInvHessProduct(s[:n_corrs], y[:n_corrs])
+
     return OptimizeResult(fun=f, jac=g, nfev=n_function_evals[0],
                           nit=n_iterations, status=warnflag, message=task_str,
-                          x=x, success=(warnflag == 0))
+                          x=x, success=(warnflag == 0), hess_inv=hess_inv)
+
+
+class _LbfgsInvHessProduct(LinearOperator):
+    """Linear operator for the BFGS approximate inverse Hessian.
+
+    References
+    ----------
+    .. [1] Nocedal, Jorge. "Updating quasi-Newton matrices with limited
+       storage." Mathematics of computation 35.151 (1980): 773-782.
+    """
+    def __init__(self, sk, yk):
+        """Construct the operator
+
+        Parameters
+        ----------
+        sk : array, shape=(n_corr, n)
+            Array of ``n_corr`` most recent updates to the solution vector.
+            (See [1]).
+        yk : array, shape=(n_corr, n)
+            Array of ``n_corr`` most recent updates to the gradient. (See [1]).
+        """
+        if sk.shape != yk.shape or sk.ndim != 2:
+            raise ValueError('sk and yk must have matching shape, (n_corrs, n)')
+        n_corrs, n = sk.shape
+
+        super(_LbfgsInvHessProduct, self).__init__(
+            dtype=np.float64, shape=(n, n))
+
+        self.sk = sk
+        self.yk = yk
+        self.n_corrs = n_corrs
+
+    def _matvec(self, x):
+        """Efficient matrix-vector multiply with the BFGS matrices.
+
+        This calculation is described in Section (4) of [1].
+
+        Parameters
+        ----------
+        x : ndarray
+            An array with shape (N,) or (N,1).
+
+        Returns
+        -------
+        y : ndarray
+            The matrix-vector product
+        """
+        x = np.asarray(x, dtype=self.dtype)
```

Comment:
maybe `q = np.array(x, dtype=self.dtype, copy=True)` since you want a copy later.